File: csharp/CsSearch/CsSearch/SearchMain.cs
using System;

namespace CsSearch
{
	class SearchMain
	{
		static void Main(string[] args)
		{
			var options = new SearchOptions();
			SearchSettings settings;
			try
			{
				settings = options.SettingsFromArgs(args);
				var searcher = new Searcher(settings);
				searcher.Search();

				if (settings.PrintResults)
				{
					Console.WriteLine();
					searcher.PrintResults();
				}

				if (settings.ListDirs)
				{
					searcher.PrintMatchingDirs();
				}

				if (settings.ListFiles)
				{
					searcher.PrintMatchingFiles();
				}

				if (settings.ListLines)
				{
					searcher.PrintMatchingLines();
				}
			}
			catch (SearchException e)
			{
				Console.WriteLine("Error: {0}", e.Message);
				options.Usage(1);
			}
		}
	}
}

File: scala/src/main/scala/scalasearch/SearchMain.scala
package scalasearch

object SearchMain {

  def log(message: String): Unit = {
    println(message)
  }

  def logError(message: String): Unit = {
    log("Error: " + message)
  }

  def main(args: Array[String]) {
    val settings =
      try {
        SearchOptions.settingsFromArgs(args)
      } catch {
        case e: SearchException =>
          log("")
          logError(e.getMessage + "\n")
          SearchOptions.usage(1)
      }

    if (settings.printUsage) {
      SearchOptions.usage(0)
    }

    try {
      val searcher = new Searcher(settings)
      searcher.search()

      if (settings.printResults) {
        log("\nSearch results (%d):".format(searcher.searchResults.length))
        searcher.printSearchResults()
      }
      if (settings.listDirs) searcher.printMatchingDirs()
      if (settings.listFiles) searcher.printMatchingFiles()
      if (settings.listLines) searcher.printMatchingLines()

    } catch {
      case e: SearchException =>
        log("")
        logError(e.getMessage + "\n")
        SearchOptions.usage(1)
    }
  }
}
